southpark-incremental-indexing/README.md

```diff
@@ -0,0 +1,163 @@
+
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**
+
+- [Running the example](#running-the-example)
+- [Overview](#overview)
+  - [What is incremental indexing?](#what-is-incremental-indexing)
+  - [Configuration changes](#configuration-changes)
+- [Documentation](#documentation)
+- [Stay tuned](#stay-tuned)
+- [License](#license)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Running the example
+
+**NOTE** Jina only works on Python 3.7 and 3.8
+
+Install requirements.
+
+`pip install -r requirements.txt`
+
+Prepare data
+
+`bash get_data.sh data`
+
+This creates all the data in the `data` dir. Of main import will be `character-lines-1.csv` and `character-lines-2.csv`.
+
+Run the indexing step
+
+`python app.py -t index`
+
+This will perform the indexing on the first file, stop the flow, re-start it, and then index the second. This showcases **incremental indexing**.
+
+Query the data
+
+`python app.py -t query`
+
+
+## Overview
+
+For the discussion on the basic topics, read through the documentation in the [South Park example](../southpark-search/README.md). This README will delve deeper into how we use **incremental indexing**.
+
+### What is incremental indexing?
+
+Briefly, this allows for the user to re-use an index to add new data. It also automatically handles adding documents with duplicate IDs. For a more in-depth explanation on how Jina handles this, refer to [this chapter](https://docs.jina.ai/chapters/prevent_duplicate_indexing/index.html) in our docs.
+
+### Configuration changes
+
+In order to adapt the South Park scripts example to support incremental indexing, we need to to the following:
+
+1. Change the `indexer` in `flow-index` and `flow-query` to use `DocIDCache` as a filter (in the `uses_before` field). This ensures that we prevent duplicates.
+
+    In this example the `DocIDCache` is separated into its own `.yml` file, in `index_cache`:
+    
+    ```yaml
+    !DocIDCache
+    with:
+      index_path: cache.tmp
+    metas:
+      name: cache
+      workspace: $JINA_WORKSPACE
+    requests:
+      on:
+        [SearchRequest, TrainRequest, IndexRequest, ControlRequest]:
+          - !RouteDriver {}
+        IndexRequest:
+          - !TaggingCacheDriver
+            with:
+              tags:
+                is_indexed: true
+          - !FilterQL
+            with:
+              lookups: {tags__is_indexed__neq: true}
+    ```
+    
+    This might look complicated, but what it basically does is to first check the cache for any matching doc IDs, before indexing and querying.
+
+2. Adapt dataset to highlight incremental indexing.
+
+    We split the dataset into two files, `character-lines-1.csv` and `characters-lines-2.csv`. This way we can show how we can index one, close the `Flow` object, and then index the other.
```

Comment:
Is there an example where we can show that the results come from the 2 different files? or that the size of the `index` is the expected at every time?

I would even add the showcase that if you want to `index` "character-lines-1.csv" a second time, nothing happens

Reply:
Good idea. It's been added in the integration test.